### PR TITLE
CMake: Set `-fvisibility-inlines-hidden` on ELF platforms

### PR DESCRIPTION
**Pull request by BertalanD**
The C++ semantics of `inline` dictate that such functions might be defined in multiple translation units. As with all other functions, they must have the same address in all TUs. Because of this, the compiler emits `inline` functions as weak symbols, which the dynamic linker can then resolve to the same address everywhere.

This rule is responsible for a significant overhead at startup, as such lookups happen by name. Namely, 86'000 of the 114'000 by-name symbol lookups when loading LibWeb can be attributed to this. Most of these are only ever defined in a single object, making this even more pointless.

As nothing in our code relies on either ELF symbol interposition rules or function address equality, we can use the -fvisibility-inlines-hidden escape hatch, which causes this rule to be disregarded. As the symbols are now hidden, no load-time symbol lookup is needed. This flag is used without issues in other large C++ codebases like Chromium and LLVM.

Some relevant light reading, suggested by Nico:
- https://ridiculousfish.com/blog/posts/i-didnt-order-that-so-why-is-it-on-my-bill-episode-1.html
- https://www.cs.dartmouth.edu/~sergey/cs258/ABI/UlrichDrepper-How-To-Write-Shared-Libraries.pdf
- https://blog.llvm.org/2018/11/30-faster-windows-builds-with-clang-cl_14.html
- https://gcc.gnu.org/onlinedocs/gcc/C_002b_002b-Dialect-Options.html#index-fvisibility-inlines-hidden

### CODE REVIEW

**Comment by awesomekling** [comment]
Seems like a solid idea :)

File: Meta/CMake/common_compile_options.cmake [hunk 1]
```diff
@@ -32,3 +32,8 @@ elseif (CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
     # FIXME: This warning seems useful but has too many false positives with GCC 13.
     add_compile_options(-Wno-dangling-reference)
 endif()
+
+if (UNIX AND NOT APPLE AND NOT ENABLE_FUZZERS)
+    add_compile_options(-fno-semantic-interposition)
+    add_compile_options(-fvisibility-inlines-hidden)
```
Comment (nico) [inline Meta/CMake/common_compile_options.cmake:38]: I think this one works on Apple too, and probably with fuzzers enabled as well.

Comment (BertalanD) [reply Meta/CMake/common_compile_options.cmake:38]: I tried it on arm64 macOS, and it has a negligible impact on codegen:
```
    FILE SIZE        VM SIZE    
 --------------  -------------- 
  +1.2%    +380  +1.2%    +380    [__TEXT]
  +0.8%    +128  +0.8%    +128    [__DATA_CONST]
  -0.0%     -16  -0.0%     -16    String Table
  -1.0%     -64  -1.0%     -64    Indirect Symbol Table
  -1.9%    -128  -1.9%    -128    __DATA_CONST,__got
  -0.0%    -380  -0.0%    -380    __TEXT,__text
  -0.2% -1.61Ki  +0.0%     +80    [__LINKEDIT]
  -0.0% -1.69Ki  [ = ]       0    TOTAL
```

Looks like it only really impacts the visibility of static local variables defined in `inline` functions. The functions themselves are already`.weak_def_can_be_hidden`/hidden visibility by default:
<details>
<summary>Comparison of symbols in libweb.dylib</summary>

```
❯ diff <(nm libweb-after.dylib| cut -d' ' -f2,3 | sort) <(nm libweb-before.dylib| cut -d' ' -f2,3 | sort)
764a765,771
> D __ZGVZN3Web3CSS15UnsetStyleValue3theEvE8instance
> D __ZGVZN3Web3CSS16RevertStyleValue3theEvE8instance
> D __ZGVZN3Web3CSS17InheritStyleValue3theEvE8instance
> D __ZGVZN3Web3CSS17InitialStyleValue3theEvE8instance
> D __ZGVZNK3Web4HTML17HTMLOutputElement4typeEvE6output
> D __ZGVZNK3Web4HTML19HTMLFieldSetElement4typeEvE8fieldset
> D __ZGVZNK3Web4HTML19HTMLTextAreaElement4typeEvE8textarea
767a775,781
> D __ZZN3Web3CSS15UnsetStyleValue3theEvE8instance
> D __ZZN3Web3CSS16RevertStyleValue3theEvE8instance
> D __ZZN3Web3CSS17InheritStyleValue3theEvE8instance
> D __ZZN3Web3CSS17InitialStyleValue3theEvE8instance
> D __ZZNK3Web4HTML17HTMLOutputElement4typeEvE6output
> D __ZZNK3Web4HTML19HTMLFieldSetElement4typeEvE8fieldset
> D __ZZNK3Web4HTML19HTMLTextAreaElement4typeEvE8textarea
4798a4813
> T __ZN2AK18is_ascii_hex_digitEj
9479a9495
> T __ZN3Web5Infra19is_ascii_whitespaceEj
22899,22905d22914
< d __ZGVZN3Web3CSS15UnsetStyleValue3theEvE8instance
< d __ZGVZN3Web3CSS16RevertStyleValue3theEvE8instance
< d __ZGVZN3Web3CSS17InheritStyleValue3theEvE8instance
< d __ZGVZN3Web3CSS17InitialStyleValue3theEvE8instance
< d __ZGVZNK3Web4HTML17HTMLOutputElement4typeEvE6output
< d __ZGVZNK3Web4HTML19HTMLFieldSetElement4typeEvE8fieldset
< d __ZGVZNK3Web4HTML19HTMLTextAreaElement4typeEvE8textarea
22909,22912d22917
< d __ZZN3Web3CSS15UnsetStyleValue3theEvE8instance
< d __ZZN3Web3CSS16RevertStyleValue3theEvE8instance
< d __ZZN3Web3CSS17InheritStyleValue3theEvE8instance
< d __ZZN3Web3CSS17InitialStyleValue3theEvE8instance
22914,22916d22918
< d __ZZNK3Web4HTML17HTMLOutputElement4typeEvE6output
< d __ZZNK3Web4HTML19HTMLFieldSetElement4typeEvE8fieldset
< d __ZZNK3Web4HTML19HTMLTextAreaElement4typeEvE8textarea
25073d25074
< t __ZN2AK18is_ascii_hex_digitEj
30860d30860
< t __ZN3Web5Infra19is_ascii_whitespaceEj

```
</details>

Comment (BertalanD) [reply Meta/CMake/common_compile_options.cmake:38]: hmm, and it's probably a *bad* idea that these singleton instance variables now get emitted in *each* object on ELF.

Comment (BertalanD) [reply Meta/CMake/common_compile_options.cmake:38]: https://reviews.llvm.org/D86881

> Make -fvisibility-inlines-hidden apply to static local variables in inline functions on Darwin

Added as a hack, still present on trunk.

**Pull request merged**